{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "blank-portal",
   "metadata": {},
   "source": [
    "# Cell-cell communication analysis\n",
    "\n",
    "## Performing differential expression (DE) analysis of the NanoString ROIs with `limma`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "veterinary-liabilities",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import matplotlib as mpl\n",
    "import scipy\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "#sc.logging.print_versions()\n",
    "sc.settings.set_figure_params(dpi=80)  # low dpi (dots per inch) yields small inline figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "logical-switzerland",
   "metadata": {},
   "outputs": [],
   "source": [
    "nanostring_data_path = '../KidneyDataset/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "radical-worst",
   "metadata": {},
   "source": [
    "## Read in and prepare NanoString data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "lonely-prison",
   "metadata": {},
   "outputs": [],
   "source": [
    "# target QC’d, filtered, and quantile normalized (Q3) target level count data\n",
    "# going to feed this into limma\n",
    "Kidney_Q3Norm_TargetCountMatrix = pd.read_csv(nanostring_data_path + 'Kidney_Q3Norm_TargetCountMatrix.txt',\n",
    "                                             sep='\\t', index_col=0)\n",
    "# ROI metadata\n",
    "Kidney_Sample_Annotations = pd.read_csv(nanostring_data_path + 'Kidney_Sample_Annotations.txt',\n",
    "                                             sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "mexican-publisher",
   "metadata": {},
   "outputs": [],
   "source": [
    "# here in Kidney_Sample_Annotations column 'region' there is no indication of whether tubule is proximal or distal\n",
    "# need to add this depending on the SegmentLabel - neg means proximal tubules,  PanCK - distal\n",
    "\n",
    "def add_exact_region(idx, table):\n",
    "    curr_SegmentLabel = table.loc[idx,'SegmentLabel']\n",
    "    if curr_SegmentLabel == 'Geometric Segment':\n",
    "        return('glomerulus')\n",
    "    elif curr_SegmentLabel == 'neg':\n",
    "        return('proximal tubule')\n",
    "    elif curr_SegmentLabel == 'PanCK':\n",
    "        return('distal tubule')\n",
    "    else:\n",
    "        return('error')\n",
    "Kidney_Sample_Annotations['idx'] = Kidney_Sample_Annotations.index\n",
    "Kidney_Sample_Annotations['region_exact'] = Kidney_Sample_Annotations['idx'].apply(lambda x: add_exact_region(x,Kidney_Sample_Annotations))\n",
    "Kidney_Sample_Annotations.set_index('SegmentDisplayName', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "individual-figure",
   "metadata": {},
   "outputs": [],
   "source": [
    "# getting rid of NaNs\n",
    "Kidney_Sample_Annotations['pathology'] = [elem if str(elem) != 'nan' else 'NA' for elem in Kidney_Sample_Annotations['pathology']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "interior-republican",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We will be comparing all abnormal ROIs to all healthy ROIs (acc to pathology)\n",
    "# and then this 1 list of DE I will artificially copy for each of the single cell clusters\n",
    "comparison_cluster = 'pathology'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "assisted-cigarette",
   "metadata": {},
   "outputs": [],
   "source": [
    "Kidney_Q3Norm_TargetCountMatrix = Kidney_Q3Norm_TargetCountMatrix.T\n",
    "# make sure indices match\n",
    "Kidney_Sample_Annotations = Kidney_Sample_Annotations.loc[Kidney_Q3Norm_TargetCountMatrix.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "hungarian-grocery",
   "metadata": {},
   "outputs": [],
   "source": [
    "# object with all features that passed QC (.raw.X)\n",
    "adata_full = anndata.AnnData(X = Kidney_Q3Norm_TargetCountMatrix,\n",
    "                             obs = Kidney_Sample_Annotations,\n",
    "                             var = pd.DataFrame(index = Kidney_Q3Norm_TargetCountMatrix.columns)\n",
    "                            )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "after-nashville",
   "metadata": {},
   "source": [
    "_____________________________________________________________________________________________________________________________________________________________"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "satisfied-fetish",
   "metadata": {},
   "source": [
    "# DE with limma\n",
    "\n",
    "- all abnormal ROIs vs all healthy ROIs according to pathology annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "necessary-moscow",
   "metadata": {},
   "outputs": [],
   "source": [
    "# discard the ROIs that are not glomeruli (aka tubules)\n",
    "adata_full = adata_full[adata_full.obs['pathology'].isin(['abnormal', 'healthy'])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "confident-casino",
   "metadata": {},
   "outputs": [],
   "source": [
    "# marker calling\n",
    "\n",
    "t = adata_full.X.T\n",
    "df = pd.DataFrame(data=t, columns= adata_full.obs.index, index=adata_full.var_names)\n",
    "\n",
    "meta_df = pd.DataFrame(data={'Cell':list(adata_full.obs.index),\n",
    "                             'cell_type':[ str(i) for i in adata_full.obs['pathology']],\n",
    "                             #'sample':[ str(i) for i in adata_full.obs['sample']]\n",
    "                            })\n",
    "meta_df.set_index('Cell', inplace=True)\n",
    "meta_df.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "dutch-situation",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.8/site-packages/rpy2/robjects/pandas2ri.py:15: FutureWarning: pandas.core.index is deprecated and will be removed in a future version.  The public classes are available in the top-level namespace.\n",
      "  from pandas.core.index import Index as PandasIndex\n"
     ]
    }
   ],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "viral-surge",
   "metadata": {},
   "outputs": [],
   "source": [
    "outpath = './'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "champion-contest",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "library(limma)\n",
    "library(edgeR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "handled-conference",
   "metadata": {},
   "outputs": [],
   "source": [
    "# because R replaces things\n",
    "meta_df['Cell'] = [elem.replace(' | ','...') for elem in meta_df['Cell']]\n",
    "meta_df['Cell'] = [elem.replace(' ','.') for elem in meta_df['Cell']]\n",
    "#meta_df.head()\n",
    "\n",
    "df.columns = [elem.replace(' | ','...') for elem in df.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "knowing-shade",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['abnormal', 'healthy'], dtype=object), array([77, 72]))"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(meta_df['cell_type'], return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "yellow-sword",
   "metadata": {},
   "outputs": [],
   "source": [
    "case = 'abnormal'\n",
    "ctrl = 'healthy'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "textile-timothy",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"abnormal\" \"healthy\" \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: 149 shared cells\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"case\" \"ctrl\"\n"
     ]
    }
   ],
   "source": [
    "%%R -i df -i meta_df -i outpath -i ctrl -i case \n",
    "\n",
    "library(limma)\n",
    "library(edgeR)\n",
    "\n",
    "# Format\n",
    "ex_mat=as.matrix(df)\n",
    "rownames(meta_df) = meta_df$Cell\n",
    "\n",
    "# subset meta\n",
    "meta_df = subset(meta_df, cell_type %in% unlist(c(ctrl, case)) )\n",
    "\n",
    "print(unique(meta_df$cell_type))\n",
    "\n",
    "# Shared cells\n",
    "shared_cells = intersect(rownames(meta_df), colnames(ex_mat))\n",
    "message(length(shared_cells), ' shared cells')\n",
    "ex_mat = ex_mat[, shared_cells]\n",
    "meta_df = meta_df[shared_cells,]\n",
    "\n",
    "# Filter lowly expressed genes\n",
    "keep = rowSums(ex_mat, na.rm=T) > 0.1\n",
    "ex_mat = ex_mat[ keep, ]\n",
    "keep = aveLogCPM(ex_mat) > 0.1\n",
    "ex_mat = ex_mat[ keep, ]\n",
    "\n",
    "# Extract celltypes\n",
    "cells = rownames(meta_df)\n",
    "celltypes = unique(meta_df$cell_type)\n",
    "covariates = meta_df$covariate\n",
    "\n",
    "# Extract cells in cluster and rest\n",
    "cells_case = rownames(subset(meta_df, cell_type == case))\n",
    "cells_ctrl = rownames(subset(meta_df, cell_type == ctrl)) # changed from control to ctrl\n",
    "\n",
    "# build cluster_type vector\n",
    "cluster_type = rep(0, length(cells))\n",
    "names(cluster_type) = cells\n",
    "cluster_type[ cells_case ] = 'case'\n",
    "cluster_type[ cells_ctrl ] = 'ctrl'\n",
    "\n",
    "print(unique(cluster_type))\n",
    "\n",
    "#design.matrix <- model.matrix(~ 0 + cluster_type + covariates)\n",
    "design.matrix <- model.matrix(~ 0 + cluster_type)\n",
    "\n",
    "# Now tell limma how do you want to compare (i.e. case vs control)\n",
    "contrast.matrix <- makeContrasts(caseVScontrol = cluster_typecase - cluster_typectrl, levels = design.matrix)\n",
    "\n",
    "# Make model and run contrasts\n",
    "fit <- lmFit(ex_mat, design.matrix)\n",
    "fit <- contrasts.fit(fit, contrast.matrix)\n",
    "fit <- eBayes(fit)\n",
    "\n",
    "# Make a dataframe containing the important data\n",
    "results = topTable(fit, adjust=\"fdr\", number = nrow(ex_mat), coef = 'caseVScontrol')\n",
    "\n",
    "# Add and filter needed data\n",
    "results$Gene = rownames(results)\n",
    "results = results[ , c('Gene', 'logFC', 'P.Value', 'adj.P.Val')]\n",
    "results$AveExpr_cluster = apply(ex_mat[ results$Gene, cells_case], 1, mean)\n",
    "results$AveExpr_rest = apply(ex_mat[ results$Gene, cells_ctrl], 1, mean)\n",
    "results$percentExpr_cluster = apply(ex_mat[ results$Gene, cells_case], 1, function(x) sum(c(x > 0)+0) ) / length(cells_case)\n",
    "results$percentExpr_rest = apply(ex_mat[ results$Gene, cells_ctrl], 1, function(x) sum(c(x > 0)+0) ) / length(cells_ctrl)\n",
    "\n",
    "results$AveExpr_cluster = round(results$AveExpr_cluster, 6)\n",
    "results$AveExpr_rest = round(results$AveExpr_rest, 6)\n",
    "results$percentExpr_cluster = round(results$percentExpr_cluster, 6)\n",
    "results$percentExpr_rest = round(results$percentExpr_rest, 6)\n",
    "# and store it as csv file\n",
    "write.csv(results, file = paste0(outpath, case, '_vs_', ctrl, '_limma_DEGs.csv'), row.names = F, col.names = T, quote = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "specialized-confirmation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>AveExpr_cluster</th>\n",
       "      <th>AveExpr_rest</th>\n",
       "      <th>percentExpr_cluster</th>\n",
       "      <th>percentExpr_rest</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HTRA1</th>\n",
       "      <td>-147.903284</td>\n",
       "      <td>3.151900e-21</td>\n",
       "      <td>5.069516e-17</td>\n",
       "      <td>160.806267</td>\n",
       "      <td>308.709550</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DDN</th>\n",
       "      <td>-29.632876</td>\n",
       "      <td>2.596122e-19</td>\n",
       "      <td>1.989300e-15</td>\n",
       "      <td>24.005380</td>\n",
       "      <td>53.638256</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AIF1</th>\n",
       "      <td>-34.851114</td>\n",
       "      <td>3.710458e-19</td>\n",
       "      <td>1.989300e-15</td>\n",
       "      <td>34.119944</td>\n",
       "      <td>68.971058</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NEDD9</th>\n",
       "      <td>-18.092348</td>\n",
       "      <td>5.381394e-19</td>\n",
       "      <td>2.163858e-15</td>\n",
       "      <td>21.228594</td>\n",
       "      <td>39.320942</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALDOB</th>\n",
       "      <td>-103.131731</td>\n",
       "      <td>2.539504e-18</td>\n",
       "      <td>8.169075e-15</td>\n",
       "      <td>32.868548</td>\n",
       "      <td>136.000279</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC26A11</th>\n",
       "      <td>0.000132</td>\n",
       "      <td>9.998341e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>9.855141</td>\n",
       "      <td>9.855009</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNPEP</th>\n",
       "      <td>0.000157</td>\n",
       "      <td>9.999368e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>38.436916</td>\n",
       "      <td>38.436759</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTC37</th>\n",
       "      <td>0.000041</td>\n",
       "      <td>9.999576e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>15.905058</td>\n",
       "      <td>15.905017</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDSN</th>\n",
       "      <td>0.000035</td>\n",
       "      <td>9.999589e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>10.358523</td>\n",
       "      <td>10.358489</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYL10</th>\n",
       "      <td>-0.000033</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>9.058847</td>\n",
       "      <td>9.058879</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16084 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               logFC       P.Value     adj.P.Val  AveExpr_cluster  \\\n",
       "Gene                                                                \n",
       "HTRA1    -147.903284  3.151900e-21  5.069516e-17       160.806267   \n",
       "DDN       -29.632876  2.596122e-19  1.989300e-15        24.005380   \n",
       "AIF1      -34.851114  3.710458e-19  1.989300e-15        34.119944   \n",
       "NEDD9     -18.092348  5.381394e-19  2.163858e-15        21.228594   \n",
       "ALDOB    -103.131731  2.539504e-18  8.169075e-15        32.868548   \n",
       "...              ...           ...           ...              ...   \n",
       "SLC26A11    0.000132  9.998341e-01  9.999668e-01         9.855141   \n",
       "RNPEP       0.000157  9.999368e-01  9.999668e-01        38.436916   \n",
       "TTC37       0.000041  9.999576e-01  9.999668e-01        15.905058   \n",
       "CDSN        0.000035  9.999589e-01  9.999668e-01        10.358523   \n",
       "MYL10      -0.000033  9.999668e-01  9.999668e-01         9.058847   \n",
       "\n",
       "          AveExpr_rest  percentExpr_cluster  percentExpr_rest  \n",
       "Gene                                                           \n",
       "HTRA1       308.709550                    1                 1  \n",
       "DDN          53.638256                    1                 1  \n",
       "AIF1         68.971058                    1                 1  \n",
       "NEDD9        39.320942                    1                 1  \n",
       "ALDOB       136.000279                    1                 1  \n",
       "...                ...                  ...               ...  \n",
       "SLC26A11      9.855009                    1                 1  \n",
       "RNPEP        38.436759                    1                 1  \n",
       "TTC37        15.905017                    1                 1  \n",
       "CDSN         10.358489                    1                 1  \n",
       "MYL10         9.058879                    1                 1  \n",
       "\n",
       "[16084 rows x 7 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# let's have a look at the DEGs\n",
    "DE_table = pd.read_csv('./abnormal_vs_healthy_limma_DEGs.csv', index_col=0)\n",
    "DE_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "sound-guyana",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>AveExpr_cluster</th>\n",
       "      <th>AveExpr_rest</th>\n",
       "      <th>percentExpr_cluster</th>\n",
       "      <th>percentExpr_rest</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HTRA1</th>\n",
       "      <td>-147.903284</td>\n",
       "      <td>3.151900e-21</td>\n",
       "      <td>5.069516e-17</td>\n",
       "      <td>160.806267</td>\n",
       "      <td>308.709550</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DDN</th>\n",
       "      <td>-29.632876</td>\n",
       "      <td>2.596122e-19</td>\n",
       "      <td>1.989300e-15</td>\n",
       "      <td>24.005380</td>\n",
       "      <td>53.638256</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AIF1</th>\n",
       "      <td>-34.851114</td>\n",
       "      <td>3.710458e-19</td>\n",
       "      <td>1.989300e-15</td>\n",
       "      <td>34.119944</td>\n",
       "      <td>68.971058</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NEDD9</th>\n",
       "      <td>-18.092348</td>\n",
       "      <td>5.381394e-19</td>\n",
       "      <td>2.163858e-15</td>\n",
       "      <td>21.228594</td>\n",
       "      <td>39.320942</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALDOB</th>\n",
       "      <td>-103.131731</td>\n",
       "      <td>2.539504e-18</td>\n",
       "      <td>8.169075e-15</td>\n",
       "      <td>32.868548</td>\n",
       "      <td>136.000279</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANGPT2</th>\n",
       "      <td>1.529483</td>\n",
       "      <td>1.117791e-02</td>\n",
       "      <td>4.989884e-02</td>\n",
       "      <td>11.128993</td>\n",
       "      <td>9.599509</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LRRC8B</th>\n",
       "      <td>1.704521</td>\n",
       "      <td>1.118756e-02</td>\n",
       "      <td>4.992562e-02</td>\n",
       "      <td>12.303394</td>\n",
       "      <td>10.598874</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR10T2</th>\n",
       "      <td>1.993797</td>\n",
       "      <td>1.119012e-02</td>\n",
       "      <td>4.992562e-02</td>\n",
       "      <td>11.400386</td>\n",
       "      <td>9.406589</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLH</th>\n",
       "      <td>1.536420</td>\n",
       "      <td>1.119361e-02</td>\n",
       "      <td>4.992735e-02</td>\n",
       "      <td>9.032784</td>\n",
       "      <td>7.496365</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIVEP2</th>\n",
       "      <td>-2.691302</td>\n",
       "      <td>1.120241e-02</td>\n",
       "      <td>4.995273e-02</td>\n",
       "      <td>20.765516</td>\n",
       "      <td>23.456818</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3607 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             logFC       P.Value     adj.P.Val  AveExpr_cluster  AveExpr_rest  \\\n",
       "Gene                                                                            \n",
       "HTRA1  -147.903284  3.151900e-21  5.069516e-17       160.806267    308.709550   \n",
       "DDN     -29.632876  2.596122e-19  1.989300e-15        24.005380     53.638256   \n",
       "AIF1    -34.851114  3.710458e-19  1.989300e-15        34.119944     68.971058   \n",
       "NEDD9   -18.092348  5.381394e-19  2.163858e-15        21.228594     39.320942   \n",
       "ALDOB  -103.131731  2.539504e-18  8.169075e-15        32.868548    136.000279   \n",
       "...            ...           ...           ...              ...           ...   \n",
       "ANGPT2    1.529483  1.117791e-02  4.989884e-02        11.128993      9.599509   \n",
       "LRRC8B    1.704521  1.118756e-02  4.992562e-02        12.303394     10.598874   \n",
       "OR10T2    1.993797  1.119012e-02  4.992562e-02        11.400386      9.406589   \n",
       "POLH      1.536420  1.119361e-02  4.992735e-02         9.032784      7.496365   \n",
       "HIVEP2   -2.691302  1.120241e-02  4.995273e-02        20.765516     23.456818   \n",
       "\n",
       "        percentExpr_cluster  percentExpr_rest  \n",
       "Gene                                           \n",
       "HTRA1                     1                 1  \n",
       "DDN                       1                 1  \n",
       "AIF1                      1                 1  \n",
       "NEDD9                     1                 1  \n",
       "ALDOB                     1                 1  \n",
       "...                     ...               ...  \n",
       "ANGPT2                    1                 1  \n",
       "LRRC8B                    1                 1  \n",
       "OR10T2                    1                 1  \n",
       "POLH                      1                 1  \n",
       "HIVEP2                    1                 1  \n",
       "\n",
       "[3607 rows x 7 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many significant DEGs?\n",
    "DE_table_significant = DE_table[DE_table['adj.P.Val'] < 0.05]\n",
    "DE_table_significant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "improving-grave",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "interpreted-directory",
   "metadata": {},
   "source": [
    "# Creating a joint DE table out of limma outputs from NanoString data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fundamental-native",
   "metadata": {},
   "source": [
    "Because we don't have cell type resolution here, let's copy this DE table for every cell type artificially to then intersect with the expressed L/R from Young et al. data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "appreciated-harrison",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['B cell',\n",
       " 'CD8 T cell',\n",
       " 'DC',\n",
       " 'Distal_tubules_and_collecting_duct',\n",
       " 'Endothelium',\n",
       " 'Loop_of_Henle',\n",
       " 'MNP',\n",
       " 'Mast cell',\n",
       " 'Mesangial_cells',\n",
       " 'NK cell',\n",
       " 'NKT cell',\n",
       " 'Nephron_epithelium',\n",
       " 'Neutrophil',\n",
       " 'Podocytes',\n",
       " 'Private',\n",
       " 'Proximal_tubular_cells',\n",
       " 'Th cell']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# from Young et al.\n",
    "meta = pd.read_csv('./cellphonedb_meta.tsv', sep='\\t')\n",
    "cell_types = np.unique(meta['cell_type'])\n",
    "# get rid of 'celltype_'\n",
    "cell_types = [elem[9:] for elem in cell_types]\n",
    "cell_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ancient-filename",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>AveExpr_cluster</th>\n",
       "      <th>AveExpr_rest</th>\n",
       "      <th>percentExpr_cluster</th>\n",
       "      <th>percentExpr_rest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HTRA1</td>\n",
       "      <td>-147.903284</td>\n",
       "      <td>3.151900e-21</td>\n",
       "      <td>5.069516e-17</td>\n",
       "      <td>160.806267</td>\n",
       "      <td>308.709550</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDN</td>\n",
       "      <td>-29.632876</td>\n",
       "      <td>2.596122e-19</td>\n",
       "      <td>1.989300e-15</td>\n",
       "      <td>24.005380</td>\n",
       "      <td>53.638256</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AIF1</td>\n",
       "      <td>-34.851114</td>\n",
       "      <td>3.710458e-19</td>\n",
       "      <td>1.989300e-15</td>\n",
       "      <td>34.119944</td>\n",
       "      <td>68.971058</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NEDD9</td>\n",
       "      <td>-18.092348</td>\n",
       "      <td>5.381394e-19</td>\n",
       "      <td>2.163858e-15</td>\n",
       "      <td>21.228594</td>\n",
       "      <td>39.320942</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALDOB</td>\n",
       "      <td>-103.131731</td>\n",
       "      <td>2.539504e-18</td>\n",
       "      <td>8.169075e-15</td>\n",
       "      <td>32.868548</td>\n",
       "      <td>136.000279</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16079</th>\n",
       "      <td>SLC26A11</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>9.998341e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>9.855141</td>\n",
       "      <td>9.855009</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16080</th>\n",
       "      <td>RNPEP</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>9.999368e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>38.436916</td>\n",
       "      <td>38.436759</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16081</th>\n",
       "      <td>TTC37</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>9.999576e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>15.905058</td>\n",
       "      <td>15.905017</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16082</th>\n",
       "      <td>CDSN</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>9.999589e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>10.358523</td>\n",
       "      <td>10.358489</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16083</th>\n",
       "      <td>MYL10</td>\n",
       "      <td>-0.000033</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>9.058847</td>\n",
       "      <td>9.058879</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16084 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene       logFC       P.Value     adj.P.Val  AveExpr_cluster  \\\n",
       "0         HTRA1 -147.903284  3.151900e-21  5.069516e-17       160.806267   \n",
       "1           DDN  -29.632876  2.596122e-19  1.989300e-15        24.005380   \n",
       "2          AIF1  -34.851114  3.710458e-19  1.989300e-15        34.119944   \n",
       "3         NEDD9  -18.092348  5.381394e-19  2.163858e-15        21.228594   \n",
       "4         ALDOB -103.131731  2.539504e-18  8.169075e-15        32.868548   \n",
       "...         ...         ...           ...           ...              ...   \n",
       "16079  SLC26A11    0.000132  9.998341e-01  9.999668e-01         9.855141   \n",
       "16080     RNPEP    0.000157  9.999368e-01  9.999668e-01        38.436916   \n",
       "16081     TTC37    0.000041  9.999576e-01  9.999668e-01        15.905058   \n",
       "16082      CDSN    0.000035  9.999589e-01  9.999668e-01        10.358523   \n",
       "16083     MYL10   -0.000033  9.999668e-01  9.999668e-01         9.058847   \n",
       "\n",
       "       AveExpr_rest  percentExpr_cluster  percentExpr_rest  \n",
       "0        308.709550                    1                 1  \n",
       "1         53.638256                    1                 1  \n",
       "2         68.971058                    1                 1  \n",
       "3         39.320942                    1                 1  \n",
       "4        136.000279                    1                 1  \n",
       "...             ...                  ...               ...  \n",
       "16079      9.855009                    1                 1  \n",
       "16080     38.436759                    1                 1  \n",
       "16081     15.905017                    1                 1  \n",
       "16082     10.358489                    1                 1  \n",
       "16083      9.058879                    1                 1  \n",
       "\n",
       "[16084 rows x 8 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# let's have a look at the DEGs\n",
    "DE_table = pd.read_csv('./abnormal_vs_healthy_limma_DEGs.csv')\n",
    "DE_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "medieval-engagement",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B cell\n",
      "CD8 T cell\n",
      "DC\n",
      "Distal_tubules_and_collecting_duct\n",
      "Endothelium\n",
      "Loop_of_Henle\n",
      "MNP\n",
      "Mast cell\n",
      "Mesangial_cells\n",
      "NK cell\n",
      "NKT cell\n",
      "Nephron_epithelium\n",
      "Neutrophil\n",
      "Podocytes\n",
      "Private\n",
      "Proximal_tubular_cells\n",
      "Th cell\n"
     ]
    }
   ],
   "source": [
    "# artificially copying it for all the cell types\n",
    "DE_tables = {}\n",
    "\n",
    "for ct in cell_types:\n",
    "    #print(ct)\n",
    "    DE_tables[ct] = DE_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "surprising-matthew",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>AveExpr_cluster</th>\n",
       "      <th>AveExpr_rest</th>\n",
       "      <th>percentExpr_cluster</th>\n",
       "      <th>percentExpr_rest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HTRA1</td>\n",
       "      <td>-147.903284</td>\n",
       "      <td>3.151900e-21</td>\n",
       "      <td>5.069516e-17</td>\n",
       "      <td>160.806267</td>\n",
       "      <td>308.709550</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDN</td>\n",
       "      <td>-29.632876</td>\n",
       "      <td>2.596122e-19</td>\n",
       "      <td>1.989300e-15</td>\n",
       "      <td>24.005380</td>\n",
       "      <td>53.638256</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AIF1</td>\n",
       "      <td>-34.851114</td>\n",
       "      <td>3.710458e-19</td>\n",
       "      <td>1.989300e-15</td>\n",
       "      <td>34.119944</td>\n",
       "      <td>68.971058</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NEDD9</td>\n",
       "      <td>-18.092348</td>\n",
       "      <td>5.381394e-19</td>\n",
       "      <td>2.163858e-15</td>\n",
       "      <td>21.228594</td>\n",
       "      <td>39.320942</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALDOB</td>\n",
       "      <td>-103.131731</td>\n",
       "      <td>2.539504e-18</td>\n",
       "      <td>8.169075e-15</td>\n",
       "      <td>32.868548</td>\n",
       "      <td>136.000279</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16079</th>\n",
       "      <td>SLC26A11</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>9.998341e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>9.855141</td>\n",
       "      <td>9.855009</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16080</th>\n",
       "      <td>RNPEP</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>9.999368e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>38.436916</td>\n",
       "      <td>38.436759</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16081</th>\n",
       "      <td>TTC37</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>9.999576e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>15.905058</td>\n",
       "      <td>15.905017</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16082</th>\n",
       "      <td>CDSN</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>9.999589e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>10.358523</td>\n",
       "      <td>10.358489</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16083</th>\n",
       "      <td>MYL10</td>\n",
       "      <td>-0.000033</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>9.058847</td>\n",
       "      <td>9.058879</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16084 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene       logFC       P.Value     adj.P.Val  AveExpr_cluster  \\\n",
       "0         HTRA1 -147.903284  3.151900e-21  5.069516e-17       160.806267   \n",
       "1           DDN  -29.632876  2.596122e-19  1.989300e-15        24.005380   \n",
       "2          AIF1  -34.851114  3.710458e-19  1.989300e-15        34.119944   \n",
       "3         NEDD9  -18.092348  5.381394e-19  2.163858e-15        21.228594   \n",
       "4         ALDOB -103.131731  2.539504e-18  8.169075e-15        32.868548   \n",
       "...         ...         ...           ...           ...              ...   \n",
       "16079  SLC26A11    0.000132  9.998341e-01  9.999668e-01         9.855141   \n",
       "16080     RNPEP    0.000157  9.999368e-01  9.999668e-01        38.436916   \n",
       "16081     TTC37    0.000041  9.999576e-01  9.999668e-01        15.905058   \n",
       "16082      CDSN    0.000035  9.999589e-01  9.999668e-01        10.358523   \n",
       "16083     MYL10   -0.000033  9.999668e-01  9.999668e-01         9.058847   \n",
       "\n",
       "       AveExpr_rest  percentExpr_cluster  percentExpr_rest  \n",
       "0        308.709550                    1                 1  \n",
       "1         53.638256                    1                 1  \n",
       "2         68.971058                    1                 1  \n",
       "3         39.320942                    1                 1  \n",
       "4        136.000279                    1                 1  \n",
       "...             ...                  ...               ...  \n",
       "16079      9.855009                    1                 1  \n",
       "16080     38.436759                    1                 1  \n",
       "16081     15.905017                    1                 1  \n",
       "16082     10.358489                    1                 1  \n",
       "16083      9.058879                    1                 1  \n",
       "\n",
       "[16084 rows x 8 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DE_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "grateful-vegetable",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B cell\n",
      "CD8 T cell\n",
      "DC\n",
      "Distal_tubules_and_collecting_duct\n",
      "Endothelium\n",
      "Loop_of_Henle\n",
      "MNP\n",
      "Mast cell\n",
      "Mesangial_cells\n",
      "NK cell\n",
      "NKT cell\n",
      "Nephron_epithelium\n",
      "Neutrophil\n",
      "Podocytes\n",
      "Private\n",
      "Proximal_tubular_cells\n",
      "Th cell\n"
     ]
    }
   ],
   "source": [
    "cluster_list = []\n",
    "\n",
    "for ct in cell_types:\n",
    "    #print(ct)\n",
    "    cluster_list.append([ct]*len(DE_table))\n",
    "\n",
    "cluster_list = [item for sublist in cluster_list for item in sublist]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "heard-reynolds",
   "metadata": {},
   "outputs": [],
   "source": [
    "# without any filtering\n",
    "joint_DE_table = pd.concat(DE_tables.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "lined-chassis",
   "metadata": {},
   "outputs": [],
   "source": [
    "joint_DE_table['cluster'] = cluster_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "gothic-reserve",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>AveExpr_cluster</th>\n",
       "      <th>AveExpr_rest</th>\n",
       "      <th>percentExpr_cluster</th>\n",
       "      <th>percentExpr_rest</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HTRA1</td>\n",
       "      <td>-147.903284</td>\n",
       "      <td>3.151900e-21</td>\n",
       "      <td>5.069516e-17</td>\n",
       "      <td>160.806267</td>\n",
       "      <td>308.709550</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDN</td>\n",
       "      <td>-29.632876</td>\n",
       "      <td>2.596122e-19</td>\n",
       "      <td>1.989300e-15</td>\n",
       "      <td>24.005380</td>\n",
       "      <td>53.638256</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AIF1</td>\n",
       "      <td>-34.851114</td>\n",
       "      <td>3.710458e-19</td>\n",
       "      <td>1.989300e-15</td>\n",
       "      <td>34.119944</td>\n",
       "      <td>68.971058</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NEDD9</td>\n",
       "      <td>-18.092348</td>\n",
       "      <td>5.381394e-19</td>\n",
       "      <td>2.163858e-15</td>\n",
       "      <td>21.228594</td>\n",
       "      <td>39.320942</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALDOB</td>\n",
       "      <td>-103.131731</td>\n",
       "      <td>2.539504e-18</td>\n",
       "      <td>8.169075e-15</td>\n",
       "      <td>32.868548</td>\n",
       "      <td>136.000279</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>B cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16079</th>\n",
       "      <td>SLC26A11</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>9.998341e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>9.855141</td>\n",
       "      <td>9.855009</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Th cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16080</th>\n",
       "      <td>RNPEP</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>9.999368e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>38.436916</td>\n",
       "      <td>38.436759</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Th cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16081</th>\n",
       "      <td>TTC37</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>9.999576e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>15.905058</td>\n",
       "      <td>15.905017</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Th cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16082</th>\n",
       "      <td>CDSN</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>9.999589e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>10.358523</td>\n",
       "      <td>10.358489</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Th cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16083</th>\n",
       "      <td>MYL10</td>\n",
       "      <td>-0.000033</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>9.999668e-01</td>\n",
       "      <td>9.058847</td>\n",
       "      <td>9.058879</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Th cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>273428 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene       logFC       P.Value     adj.P.Val  AveExpr_cluster  \\\n",
       "0         HTRA1 -147.903284  3.151900e-21  5.069516e-17       160.806267   \n",
       "1           DDN  -29.632876  2.596122e-19  1.989300e-15        24.005380   \n",
       "2          AIF1  -34.851114  3.710458e-19  1.989300e-15        34.119944   \n",
       "3         NEDD9  -18.092348  5.381394e-19  2.163858e-15        21.228594   \n",
       "4         ALDOB -103.131731  2.539504e-18  8.169075e-15        32.868548   \n",
       "...         ...         ...           ...           ...              ...   \n",
       "16079  SLC26A11    0.000132  9.998341e-01  9.999668e-01         9.855141   \n",
       "16080     RNPEP    0.000157  9.999368e-01  9.999668e-01        38.436916   \n",
       "16081     TTC37    0.000041  9.999576e-01  9.999668e-01        15.905058   \n",
       "16082      CDSN    0.000035  9.999589e-01  9.999668e-01        10.358523   \n",
       "16083     MYL10   -0.000033  9.999668e-01  9.999668e-01         9.058847   \n",
       "\n",
       "       AveExpr_rest  percentExpr_cluster  percentExpr_rest  cluster  \n",
       "0        308.709550                    1                 1   B cell  \n",
       "1         53.638256                    1                 1   B cell  \n",
       "2         68.971058                    1                 1   B cell  \n",
       "3         39.320942                    1                 1   B cell  \n",
       "4        136.000279                    1                 1   B cell  \n",
       "...             ...                  ...               ...      ...  \n",
       "16079      9.855009                    1                 1  Th cell  \n",
       "16080     38.436759                    1                 1  Th cell  \n",
       "16081     15.905017                    1                 1  Th cell  \n",
       "16082     10.358489                    1                 1  Th cell  \n",
       "16083      9.058879                    1                 1  Th cell  \n",
       "\n",
       "[273428 rows x 9 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joint_DE_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "productive-courage",
   "metadata": {},
   "outputs": [],
   "source": [
    "joint_DE_table.to_csv('./joint_DEGs_list_all_cell_types_for_cellphone.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hungry-tobago",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "municipal-trademark",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "transparent-grammar",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Nanostring Hack 2021",
   "language": "python",
   "name": "nanostring-hack-2021"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
